quadratichq/quadratic
# feat: education enrollment

## How it Works

- A new piece of data is added to the User table in the database: `edu_status`
- It’s an enum whose values are `null`, `INELIGIBLE`, ~~`ELIGIBLE`~~ and `ENROLLED`
- There’s a new endpoint in the API at `/education`.
  - A `GET` returns the current value of `edu_status` from the database.
  - A `POST` with `refresh` will check whether the requesting user's email is a valid education email (by looking at [this list](https://github.com/Hipo/university-domains-list) and checking our whitelist of domains in sanity). If it is, it switches the status to `ENROLLED` which signals to the front-end to display a little education badge in the bottom corner.
- A `POST` with `eduStatus` will update the current value (not used currently)
  
Enrolled: 

![CleanShot 2024-04-29 at 21 12 30@2x](https://github.com/quadratichq/quadratic/assets/1316441/52b16dbd-0b9b-4fe1-ab50-61ebb5ec772d)

Also in the sidebar (it's clickable and shows the dialog above):

![CleanShot 2024-04-29 at 21 13 39@2x](https://github.com/quadratichq/quadratic/assets/1316441/361f5475-6b4a-4148-89f1-7e938eebf99b)

Ineligible:

![CleanShot 2024-04-29 at 21 13 00@2x](https://github.com/quadratichq/quadratic/assets/1316441/b02a0ca3-5c8f-4833-8fad-6aa73d50fb7c)


## To Test

- Login to the preview branch and hit the `/education` endpoint
- See that you are ineligible
- Login to [sanity CMS](https://quadratic-cms.vercel.app) and update the whitelist to include the domain of your current account's email
- Hit the `/education` endpoint again
- See that you are now "Enrolled" by the status in the lower left of the sidebar (and clicking on it reveals the dialog with a "Learn more" link to the docs page)

## Notes

- If at any point you want to change your status from ENROLLED back to INELIGIBLE, you can either 1) go into the database and change the field yourself, or 2) remove your email address from the edu whitelist (if it's there) and hit the `/education` route again
- In the future, additional states could be possible like `ELIGIBLE` that would allow the user to explicitly opt-in/out of the features. For now, we opt you in if you're considered eligible.
- Manually changing someone's status in the database from `INELIGIBLE` to `ENROLLED` _will not_ show them the education dialog when they log in. They only way they will see the education dialog is either by: 1) visiting `/education`, or 2) if they're eligible, clicking the "Education" item in the sidebar.

## TODOs

- [x] Decide on what to do with the schema
- [x] Write API tests for `/education` endpoints

## Before/After Merging

- [x] [BEFORE] Need to review/merge https://github.com/quadratichq/quadratic-cms/pull/1/files
- [ ] [AFTER] Publish education docs
- [ ] [AFTER] Update education page on website

Resolves #1272
